Add initial support for buffered atomics

ronawho opened the pull request:
This adds a package module that supports buffered network atomics for
non-fetching operations. It includes `addBuff()`, `subBuff()`, `orBuff()`,
`andBuff()`, and `xorBuff()`. These buffered atomics are not consistent with
regular atomic operations and currently must be synchronized manually with a
call to `flushAtomicBuff()`. e.g.

```chpl
var a: atomic int;
a.addBuff(1);
writeln(a);        // prints 0
flushAtomicBuff();
writeln(a);        // prints 1
```

Under the covers these buffered atomics are implemented with 64 element
thread-local buffers. When full, these buffers are flushed. We use chained
transactions to initiate the network atomics, which offers significantly
increased transaction rate.

For the bale histogram benchmark this improves performance by over 4x.

Currently the code is written like:

```chpl
forall r in rindex do
  A[r].add(1);
```

And with buffered atomics looks something like:

```chpl
use BufferedAtomics;
forall r in rindex do
  A[r].addBuff(1);
flushAtomicBuff();
```

On a 28-core broadwell system the normal atomic implementation gets 80 MB/s per
node and the buffered atomic implementation achieves 360 MB/s per node, which
is on par with the non-blocking UPC implementation.

Using buffered atomics significantly improve ra-atomics as well and we see a
similar ~4.5x speedup. Scalability results with 256M hugepages up to 256 nodes
show below:

![ra-perf](https://user-images.githubusercontent.com/1588337/43859638-338a7d52-9b1f-11e8-8b72-cec2ffb65c40.png)


Closes https://github.com/chapel-lang/chapel/issues/10495

ronawho commented on runtime/src/comm/ugni/comm-ugni.c:
```diff
@@ -752,11 +752,12 @@ static size_t rdma_threshold = DEFAULT_RDMA_THRESHOLD;
+// this at once didn't seem to improve performance.
 //
 #define MAX_CHAINED_PUT_LEN 64
+#define MAX_CHAINED_AMO_LEN 64
```
FYI I did tuning on this number and did not just copy the put number. I see max performance with anything above 32, and saw degraded performance with 128 or higher.

ronawho commented on runtime/src/comm/ugni/comm-ugni.c:
```diff
@@ -6710,6 +6795,123 @@ int amo_cmd_2_nic_op(fork_amo_cmd_t cmd, int fetching)
+    CHPL_INTERNAL_ERROR("do_nic_amo(): "
+                        "remote address is not NIC-registered");
+
+  while (atomic_exchange_bool(&lock, true)) { local_yield(); }
```
I had been worried about the cost of this lock for each operation, but it ended up being pretty low overhead since in the normal case it's never contended. I see around 365 MB/s  for bale-histo without this lock and 360 MB/s with it.

ronawho replied:
And note that the lock is needed for 2 reasons:
 - so flushing can be done concurrently
 - so we don't run into issues with a second task trying to use the buffer when we're yielding in `post_fma_ct_and_wait`

gbtitus commented on runtime/src/comm/ugni/comm-ugni.c:
```diff
@@ -6499,6 +6505,38 @@ DEFINE_CHPL_COMM_ATOMIC_CMPXCHG(real64, cmpxchg_64, int_least64_t)
+        {                                                               \
+          mem_region_t* remote_mr;                                      \
+          DBG_P_LP(DBGF_IFACE|DBGF_AMO,                                 \
+                   "IFACE chpl_comm_atomic_"#_o"_"#_f                   \
```
... "_buff_"#_f

gbtitus commented on runtime/src/comm/ugni/comm-ugni.c:
```diff
@@ -6601,6 +6639,37 @@ DEFINE_CHPL_COMM_ATOMIC_INT_OP(uint64, add, add_i64, uint_least64_t)
+        {                                                               \
+          mem_region_t* remote_mr;                                      \
+          DBG_P_LP(DBGF_IFACE|DBGF_AMO,                                 \
+                   "IFACE chpl_comm_atomic_add_"#_f                     \
```
... `_add_buff_"`

gbtitus commented on runtime/src/comm/ugni/comm-ugni.c:
```diff
@@ -6658,6 +6727,22 @@ DEFINE_CHPL_COMM_ATOMIC_REAL_OP(real64, add_r64, double)
+          _t nopnd = - *(_t*) opnd;                                     \
+                                                                        \
+          DBG_P_LP(DBGF_IFACE|DBGF_AMO,                                 \
+                   "IFACE chpl_comm_atomic_sub_"#_f                     \
```
... `_sub_buff_"`

gbtitus commented on runtime/src/comm/ugni/comm-ugni.c:
```diff
@@ -6710,6 +6795,123 @@ int amo_cmd_2_nic_op(fork_amo_cmd_t cmd, int fetching)
+  if (inited == false) {
+    uint32_t idx;
+    while (atomic_exchange_bool(&amo_init_lock, true)) { local_yield(); }
+    atomic_init_bool(&lock, false);
```
If you want to take a teensy bit of overhead out, `lock` can be inited outside the single-threaded region because it's private to this thread.

gbtitus commented on runtime/src/comm/ugni/comm-ugni.c:
```diff
@@ -6710,6 +6795,123 @@ int amo_cmd_2_nic_op(fork_amo_cmd_t cmd, int fetching)
 }
 
+#if HAVE_GNI_FMA_CHAIN_TRANSACTIONS
+#define MAX_BUFF_AMO_THREADS 1024
```
This aspect of the design troubles me.  These are statically sized and have the potential to get used up during execution.  (The likely way that would happen is probably not long-lived tasking layer pthreads because we won't ever create 1k of those even with tasks=fifo, but rather other pthreads from other programming models or the user, coming in here from non-Chapel code calling Chapel code in a multi-language program.  Once an entry is used here it's never freed, so a program that creates a pthread which calls Chapel code, ends up grabbing an entry here, and then terminates will leave the entry dangling and unusable.)  At the very least we should emit a warning when they do get used up, since we're going to fall off a performance cliff for super-relaxed AMOs at that point.  As an alternative design that doesn't have static sizing, how about defining a thread-private struct containing all the meta-information for each thread's buffered AMOs, and then just chain those together on a (doubly) linked list, under protection of a lock?  Adding the thread-private struct to the list could be done under an `inited` flag in `do_nic_amo_nf_buff()` where you do initialization now.  You'd also need to create a pthread cleanup function (see `othread_cleanup_push()`) to remove the thread's struct from the chain if/when it terminated.

ronawho replied:
Offline we agreed to emit a warning for now, and we can address this later -- practically speaking this is unlikely to bite us in the near future (at least for Chapel-only programs.)

gbtitus commented on runtime/src/comm/ugni/comm-ugni.c:
```diff
@@ -6710,6 +6795,123 @@ int amo_cmd_2_nic_op(fork_amo_cmd_t cmd, int fetching)
+    atomic_init_bool(&lock, false);
+    idx = atomic_load_uint_least32_t(&amo_pdc_sz);
+    if (idx >= MAX_BUFF_AMO_THREADS) {
+      do_nic_amo_nf(opnd1, locale, object, size, cmd, remote_mr);
```
Don't we need to set `inited`, clear the lock, and return after this?  If we continue on we over-index `amo_pdc[]` and the rest (not to mention that the AMO will have already been done by this call and therefore shouldn't be buffered anyway).

ronawho replied:
Oh shoot, yes this is missing the early return and init_lock unlock

gbtitus commented on runtime/src/comm/ugni/comm-ugni.c:
```diff
@@ -7679,17 +7881,32 @@ int post_fma_ct(c_nodeid_t* locale_v, gni_post_descriptor_t* post_desc)
+           pdc = pdc->next_descr, i++) {
+        pdc->ep_hndl = cd->remote_eps[locale_v[i]];
+        if (post_desc->type == GNI_POST_FMA_PUT)
+          PERFSTATS_ADD(sent_bytes, pdc->length);
```
I think you can just remove the if-stmt here and do the `PERFSTATS_ADD(sent_bytes, ...)` only.  This was originally here to cover the case where we had chained GETs as well as PUTs, but we don't have chained GETs now and if we add them in the future we can update this.

gbtitus commented on runtime/src/comm/ugni/comm-ugni.c:
```diff
@@ -7679,17 +7881,32 @@ int post_fma_ct(c_nodeid_t* locale_v, gni_post_descriptor_t* post_desc)
+           pdc != NULL;
+           pdc = pdc->next_descr, i++) {
+        pdc->ep_hndl = cd->remote_eps[locale_v[i]];
+        if (post_desc->type == GNI_POST_FMA_PUT)
```
This whole if-stmt can be removed.

ronawho commented:
Rebased, fixed the logic for when we run out of table entries (and added a warning), fixed the debug prints, and renamed `atomicBuffFlush()` to `flushAtomicBuff()`

gbtitus commented on runtime/src/comm/ugni/comm-ugni.c:
```diff
@@ -7690,15 +7902,24 @@ int post_fma_ct(c_nodeid_t* locale_v, gni_post_descriptor_t* post_desc)
          pdc = pdc->next_descr, i++) {
       pdc->ep_hndl = cd->remote_eps[locale_v[i]];
-      PERFSTATS_ADD_POST(post_desc);
+      PERFSTATS_ADD(sent_bytes, post_desc->length);
```
`pdc->length`?

ronawho commented:
> Looks like there's no documentation for this yet (?). Can we add it by the release?

Correct, and yes I plan on adding it (it's the first thing listed in https://github.com/chapel-lang/chapel/issues/10743)

ronawho commented:
>> Looks like there's no documentation for this yet (?). Can we add it by the release?

> Correct, and yes I plan on adding it (it's the first thing listed in #10743)

Initial docs added in https://github.com/chapel-lang/chapel/pull/10900 and are available at https://chapel-lang.org/docs/master/modules/packages/BufferedAtomics.html. They've been reviewed by Greg and Bharsh so far, but I'll probably send them out for general group feedback too
